### DISCUSSION TITLE
How can I forcibly remove a bucket with a retention policy ?

### DISCUSSION THREAD

**Discussion by baofengzhouyu** [minio/minio#15373] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
I got an error when i delete a bucket with a retention policy.

mc rb minio/tmp2  --force
mc: <ERROR> Failed to remove `minio/tmp2`. Object, 'tiflash.zip (Version ID=97141b9d-3078-47ed-afe3-174643514122)' is WORM protected and cannot be overwritten.

**Describe the solution you'd like**
I want to remove this bucket tmp2 without retaining any data.

**Describe alternatives you've considered**
I try to do the following ,i got the same error.

mc retention clear minio/tmp2/ -r
Unable to clear object retention on `minio/tmp2/tiflash.zip`: Object is WORM protected and cannot be overwritten.
mc: <ERROR> Invalid URL Object is WORM protected and cannot be overwritten
mc: <ERROR> Unable to find any object/version to clear its retention.

**Additional context**
![image](https://user-images.githubusercontent.com/49833028/180182080-a41d50d4-87e5-44e8-ada4-d9767ab5fa98.png)

**Comment by polaris-phecda** [comment, marked as answer]
The retention policy you set is Governance mode  or Compliance mode？

- In compliance mode, a protected object version can't be overwritten or deleted by any user including the root user in your AWS account.
- In governance mode, users can't overwrite or delete an object version or alter its lock settings unless they have special  permissions. If you try to delete objects protected by governance mode and have s3:BypassGovernanceRetention  or s3:GetBucketObjectLockConfiguration permissions, the operation will succeed.

tips：
[object-lock-retention-modes](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock-overview.html#object-lock-retention-modes)
I found these content at [bucket-object-lock.go](https://github.com/minio/minio/blob/master/cmd/bucket-object-lock.go)

**Comment by baofengzhouyu** [comment]
Thank you for your reply.
 I've set both modes，I forgot that one was earlier. how can i get the modes config ?
 My user permission is s3:*. 
![image](https://user-images.githubusercontent.com/49833028/180197061-11265406-68e3-4d31-a05c-efc9e3cff53e.png)

**Reply by polaris-phecda** [reply]
`mc retention info -r --versions {alias}/tmp2` maybe help you
docs: 
[mc-retention-info](https://docs.min.io/minio/baremetal/reference/minio-mc/mc-retention-info.html)

**Comment by Rocking80** [comment]
I have created a policy which has `s3:BypassGovernanceRetention` and  `s3:GetBucketObjectLockConfiguration` enabled. Then assigned to a user `superuser`. Then created an alias to delete the object but still no lucky.

```
[root@host]# mc retention info -r --versions miniozac/demobucket
[    NO RETENTION    ]  5a6cf6bf-bd89-4dd3-9f3c-9253c8f13083  minio/demobucket/file.tgz

[root@host]# mc rm minio/demobucket/file.tgz
Creating delete marker `minio/demobucket/file.tgz` (versionId=72d71617-87e9-45a8-be56-04e686880984).

[root@host]# mc ls --versions minio/demobucket
[2022-11-09 17:53:00 CST] 141MiB STANDARD 5a6cf6bf-bd89-4dd3-9f3c-9253c8f13083 v1 PUT file.tgz

[root@host]# mc rb minio/demobucket --force 
mc: <ERROR> Failed to remove `minio/demobucket`. Object, 'file.tgz (Version ID=5a6cf6bf-bd89-4dd3-9f3c-9253c8f13083)' is WORM protected and cannot be overwritten.
```

**Comment by praveenprem** [comment]
I'm having the same issue, I've tried everything I can find but unfortunately keep running into WORM protection error.
```
mc: <ERROR> Unable to set retention on `minio/terraform-state/ubuntu/terraform.tfstate`. Object is WORM protected and cannot be overwritten.
```

My only solution was to destroy the instance completely and rebuild it from scratch. It would be ideal if there were a warning that enabling retention is irreversible.
